sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp

```diff
@@ -530,211 +376,92 @@ class wi_element<uint16_t, NumRows, NumCols, Layout, Group> {
-             const uint16_t &rhs) {
-#ifdef __SYCL_DEVICE_ONLY__
-    return make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) >=
-           make_fp32(rhs);
-#else
-    (void)lhs;
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
-#endif // __SYCL_DEVICE_ONLY__
-  }
-
-  friend bool
-  operator==(const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,
-             const uint16_t &rhs) {
-#ifdef __SYCL_DEVICE_ONLY__
-    return std::fabs(
-               make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) -
-               make_fp32(rhs)) < std::numeric_limits<float>::epsilon();
-#else
-    (void)lhs;
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
-#endif // __SYCL_DEVICE_ONLY__
-  }
+#if __SYCL_DEVICE_ONLY__
+#define OP(op)                                                                 \
+  wi_element &operator op##=(const uint16_t &rhs) {                            \
+    M.spvm = __spirv_VectorInsertDynamic(                                      \
+        M.spvm,                                                                \
+        make_bf16(make_fp32(__spirv_VectorExtractDynamic(M.spvm, idx)          \
+                                op make_fp32(rhs))),                           \
+        idx);                                                                  \
+    return *this;                                                              \
+  }
+#else // __SYCL_DEVICE_ONLY__
+#define OP(op)                                                                 \
+  wi_element &operator op##=(const uint16_t &rhs) {                            \
+    (void)rhs;                                                                 \
+    throw runtime_error("joint matrix is not supported on host device.",       \
+                        PI_INVALID_DEVICE);                                    \
+  }
+#endif // __SYCL_DEVICE_ONLY__
+  OP(+)
+  OP(-)
+  OP(*)
+  OP(/)
+#undef OP
+
+  template <typename T1, typename T2> struct Converter {
+    static T2 convert(const T1 &from) { return static_cast<T2>(from); }
+  };
 
-  friend bool
-  operator!=(const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,
-             const uint16_t &rhs) {
-#ifdef __SYCL_DEVICE_ONLY__
-    return std::fabs(
-               make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) -
-               make_fp32(rhs)) >= std::numeric_limits<float>::epsilon();
-#else
-    (void)lhs;
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
+  template <typename T> struct Converter<T, uint16_t> {
+    static uint16_t convert(const T &from) { return make_bf16(from); }
+  };
+#if __SYCL_DEVICE_ONLY__
+#define OP(input_type, type, op)                                               \
```

Comment:
Eh, I am not gifted at naming, do we have a better naming style for 'inuput_type' here? actually it represents input type of 'convert' function.